internal/app/machined/internal/mount/mount.go

```diff
@@ -0,0 +1,198 @@
+func (i *Initializer) Owned() *mount.Points {
+	return i.owned
+}
+
+// InitOwned initializes and mounts the OS owned block devices in the early boot
+// tasks.
+func (i *Initializer) InitOwned() (err error) {
+	var owned *mount.Points
+	if owned, err = mountpoints(); err != nil {
+		return errors.Errorf("error initializing owned block devices: %v", err)
+	}
+	i.owned = owned
+	if mountpoint, ok := i.owned.Get(constants.DataPartitionLabel); ok {
+		if err = repair(mountpoint); err != nil {
+			return errors.Errorf("error fixing data partition: %v", err)
+		}
+	}
+
+	iter := i.owned.Iter()
+	for iter.Next() {
+		if err = mount.WithRetry(iter.Value(), mount.WithPrefix(i.prefix)); err != nil {
+			return errors.Errorf("error mounting partitions: %v", err)
+		}
+	}
+	if iter.Err() != nil {
+		return iter.Err()
+	}
+
+	if mountpoint, ok := i.owned.Get(constants.DataPartitionLabel); ok {
+		// NB: The XFS partition MUST be mounted, or this will fail.
+		log.Printf("growing the %s partition", constants.DataPartitionLabel)
+		if err = xfs.GrowFS(path.Join(i.prefix, mountpoint.Target())); err != nil {
+			return errors.Errorf("error growing data partition file system: %v", err)
+		}
+	}
+
+	return nil
+}
+
+// ExtraDevices mounts the extra devices.
+func ExtraDevices(data *userdata.UserData) (err error) {
+	if data.Install == nil || data.Install.ExtraDevices == nil {
+		return nil
+	}
+	for _, extra := range data.Install.ExtraDevices {
+		for i, part := range extra.Partitions {
+			devname := fmt.Sprintf("%s%d", extra.Device, i+1)
+			mountpoint := mount.NewMountPoint(devname, part.MountPoint, "xfs", unix.MS_NOATIME, "")
+			if err = mount.WithRetry(mountpoint); err != nil {
+				return errors.Errorf("failed to mount %s at %s: %v", devname, part.MountPoint, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// MountOwned mounts the OS owned block devices.
+func (i *Initializer) MountOwned() (err error) {
+	iter := i.owned.Iter()
+	for iter.Next() {
+		if err = mount.WithRetry(iter.Value(), mount.WithPrefix(i.prefix)); err != nil {
+			return errors.Errorf("error mounting partitions: %v", err)
+		}
+	}
+	if iter.Err() != nil {
+		return iter.Err()
+	}
+
+	return nil
+}
+
+// UnmountOwned unmounts the OS owned block devices.
+func (i *Initializer) UnmountOwned() (err error) {
+	iter := i.owned.IterRev()
+	for iter.Next() {
+		if err = mount.UnWithRetry(iter.Value(), mount.WithPrefix(i.prefix)); err != nil {
+			return errors.Errorf("error mounting partitions: %v", err)
+		}
+	}
+	if iter.Err() != nil {
+		return iter.Err()
+	}
+
+	return nil
+}
+
+// nolint: dupl
+func mountpoints() (mountpoints *mount.Points, err error) {
+	mountpoints = mount.NewMountPoints()
+	for _, name := range []string{constants.DataPartitionLabel, constants.BootPartitionLabel} {
+		var target string
+		switch name {
+		case constants.DataPartitionLabel:
+			target = constants.DataMountPoint
+		case constants.BootPartitionLabel:
+			target = constants.BootMountPoint
```

Comment:
probably overkill, but should we add a `default: to return error` ?

Reply:
I have some local changes to clean up all the mount logic, was planning on a follow-up PR. I will take a closer look at this then.